A1-http-connect-proxy-support.md

```diff
@@ -1,222 +1,265 @@
+servers, which has security implications (e.g., a successful attack on the proxy
+would yield priviledged access on all internal servers).
 
 ## Implementation
 
-In C-core, the HTTP CONNECT handshaker was already implemented.
-The following additional changes have also been made:
+### C-Core
+
+In C-core, the HTTP CONNECT handshaker was already implemented. The following
+additional changes have also been made:
+
+-   [grpc/grpc#9383](https://github.com/grpc/grpc/pull/9383) (changes HTTP
+    CONNECT handshaker to be triggered via channel args)
+-   [grpc/grpc#9372](https://github.com/grpc/grpc/pull/9372) (adds proxy mapper
+    hook for address rewriting)
+-   [grpc/grpc#9557](https://github.com/grpc/grpc/pull/9557) (adds proxy mapper
+    hook for name rewriting)
+
+This section details C-Core's implementation of the proxy mapper interface
+available by default (Case 1).
 
-- [grpc/grpc#9383](https://github.com/grpc/grpc/pull/9383) (changes
-  HTTP CONNECT handshaker to be triggered via channel args)
-- [grpc/grpc#9372](https://github.com/grpc/grpc/pull/9372) (adds proxy
-  mapper hook for address rewriting)
-- [grpc/grpc#9557](https://github.com/grpc/grpc/pull/9557) (adds proxy
-  mapper hook for name rewriting)
+C-Core checks the following places to determine the HTTP proxy to use, stopping
+at the first one that is set:
+
+1.  `GRPC_ARG_HTTP_PROXY` channel arg
+2.  `grpc_proxy` environment variable
+3.  `https_proxy` environment variable
+4.  `http_proxy` environment variable
+
+If none of the above are set, then no HTTP proxy will be used.
+
+The allowed format is an RFC3986 URI string where the scheme is expected to be
+"http" and the authority portion is used to determine the proxy to be used. For
+example, for an HTTP proxy setting of
+`http://username:password@proxy.google.com:443`, `username:password` would be
+used as user credentials for proxy authentication as per RFC 7617 and
+`proxy.google.com:443` would be the host:port HTTP proxy target. If the port
+part of the authority is omitted, a default port of 443 is used. Note that user
+credential can also be omitted if the proxy does not need authentication.
+
+If an HTTP proxy is set, C-Core then checks the following places to exclude
+traffic destined to listed hosts from going through the proxy determined above,
+again stopping at the first one that is set:
+
+1.  `no_grpc_proxy` environment variable
+2.  `no_proxy`environment variable
+
+If none of the above are set, then the previously found HTTP proxy is used.
+
+The format takes a comma-separated list of names, and if any of these names
+matches as a suffix of the server host (provided in the channel target), then
+the proxy will not be used for that target. For example, with a `grpc_proxy`
+setting of `proxy.google.com` and a `no_grpc_proxy` setting of `example.com,
+google.com`, channel targets such as `dns:///foo.google.com:50051` and
+`bar.example.com:1234` will not use the proxy, but `baz.googleapis.com:443`
```

Comment:
Thanks for the additional description and the example. Is it possible to say something about IP addresses? For example, if the `no_proxy` list contains IP addresses will that work in any way? And the example is using `dns:///` scheme but if one were to use `xds:///` scheme how will this feature behave or this feature just cannot be used? This is mainly for my information but if we have that info it might be useful to document here as well.

Reply:
So it all depends on what host:port was mentioned when creating the channel. We don't do anything special for `xds` and still use the path mentioned in the channel target. The only way IP addresses would be useful is if the channel target specified an IP address. Let me know if I should some more examples.

Note that I'm not saying that this is how it *should* work. I'm just documenting how it works right now. We could very well change the behavior for `xds` targets.

Reply:
Correct. We just want to describe how it currently works - nothing more. Example or not, it might be worth mentioning that if the user includes an IP address in the `no_proxy` list then the only way it will be excluded is when used in the channel creation target string (and not if the target string resolves to that IP address).

Reply:
Addressed this in https://github.com/grpc/grpc/pull/32789